### PR TITLE
Chat controls

### PR DESCRIPTION
**Pull request by RedFlames**
- `Left`/`Right` arrow keys move cursor
 - `Ctrl-Left`/`Ctrl-Right` moves cursor between words

 - `Backspace` works at cursor of course
 - `Ctrl-Backspace` removes entire word

 - `Delete` key works as one would expect
 - `Ctrl-Delete` removes entire word

there's also some extra logic so that holding down a direction has an initial delay before continuous movement

### CODE REVIEW

File: CelesteNet.Client/Components/CelesteNetChatComponent.cs [hunk 1]
```diff
@@ -157,10 +177,28 @@ public CelesteNetChatComponent(CelesteNetClientContext context, Game game)
 
             if (!Active && !isRebinding && Settings.ButtonChat.Button.Pressed) {
                 Active = true;
-
             } else if (Active) {
                 Engine.Commands.Open = false;
 
+                _ControlHeld = MInput.Keyboard.Check(Keys.LeftControl) || MInput.Keyboard.Check(Keys.RightControl);
+
+                if(!MInput.Keyboard.Check(Keys.Left) && !MInput.Keyboard.Check(Keys.Right)) {
```
Comment (0x0ade) [inline CelesteNet.Client/Components/CelesteNetChatComponent.cs:185]: Please change the spacing here to match the rest of the project 😅 

```suggestion
                if (!MInput.Keyboard.Check(Keys.Left) && !MInput.Keyboard.Check(Keys.Right)) {
```

File: CelesteNet.Client/Components/CelesteNetChatComponent.cs [hunk 2]
```diff
@@ -198,19 +264,59 @@ public CelesteNetChatComponent(CelesteNetClientContext context, Game game)
...
                 _RepeatIndex = 0;
                 _Time = 0;
-
-            } else if (c == (char) 127) {
-                // Delete - currenly not handled.
-
             } else if (!char.IsControl(c)) {
-                // Any other character - append.
-                Typing += c;
+                if(CursorIndex == Typing.Length) {
```
Comment (0x0ade) [inline CelesteNet.Client/Components/CelesteNetChatComponent.cs:312]: Please change the spacing here to match the rest of the project 😅 

```suggestion
                if (CursorIndex == Typing.Length) {
```

File: CelesteNet.Client/Components/CelesteNetChatComponent.cs [hunk 3]
```diff
@@ -242,14 +348,29 @@ public CelesteNetChatComponent(CelesteNetClientContext context, Game game)
...
-                    offs += CelesteNetClientFont.Measure(text).X * scale;
-                    CelesteNetClientFont.Draw(
-                        "_",
-                        new(50f * scale + offs, UI_HEIGHT - 105f * scale),
-                        Vector2.Zero,
-                        fontScale,
-                        Color.White * 0.5f
-                    );
+
+                    if(CursorIndex == Typing.Length) {
```
Comment (0x0ade) [inline CelesteNet.Client/Components/CelesteNetChatComponent.cs:352]: Please change the spacing here to match the rest of the project 😅 

```suggestion
                    if (CursorIndex == Typing.Length) {
```

**Review by 0x0ade — CHANGES_REQUESTED** [review]
Sorry for the late response, and thank you for this pull request :) It looks good to me apart from the spacing mismatches.

I don't know if I would've gone for the `CursorIndex` setter to default to wrapping, would've probably gone for clamping instead 😅 I'm fine with it as is though.

**Comment by RedFlames** [comment]
Hey, glad to hear there are no major objections :)

> I don't know if I would've gone for the `CursorIndex` setter to default to wrapping, would've probably gone for clamping instead 😅 I'm fine with it as is though.

I'm 100% certain if the cursor actually ever wrapped around I'd go crazy 😬 I think I just took the modulo clamping from elsewhere to make sure the index never goes out of bounds for the input string. It looks like in all the places it _could_ wrap in theory, like pressing Left/Right I put extra comparisons against Typing.Length to prevent any wrapping 😅 
So I suppose I can also make the setter clamp instead of wrap.

I'll fix the mismatching whitespace asap, no problem.

**Review by 0x0ade — APPROVED** [review]
Thank you for fixing the whitespace issues, changing the `CursorIndex` setter behavior and this PR in general :)

**Pull request merged**